### PR TITLE
SONARJAVA-2416 rewrite rule to raise issue only on unread reassigned parameters

### PR DESCRIPTION
**Pull request by benzonico**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/ParameterReassignedToCheck.java [hunk 1]
```diff
@@ -43,60 +46,83 @@
...
   @Override
   public void visitMethod(MethodTree tree) {
+    BlockTree block = tree.block();
+    if(block == null) {
+      return;
+    }
+    CFG cfg = CFG.build(tree);
+    LiveVariables analyze = LiveVariables.analyze(cfg);
+    Set<Symbol> live = analyze.getIn(cfg.entry());
     for (VariableTree parameterTree : tree.parameters()) {
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/ParameterReassignedToCheck.java]: Why not rely on stream?
```
tree.parameters().stream()
  .map(VariableTree::symbol)
  .filter(symbol -> !live.contains(symbol))
  .forEach(variables::add)
```

File: java-checks/src/main/java/org/sonar/java/checks/ParameterReassignedToCheck.java [hunk 2]
```diff
@@ -43,60 +46,83 @@
...
-      variables.add(parameterTree.symbol());
+      if(!live.contains(parameterTree.symbol())) {
+        variables.add(parameterTree.symbol());
+      }
     }
     super.visitMethod(tree);
     for (VariableTree parameterTree : tree.parameters()) {
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/ParameterReassignedToCheck.java]: same comment about usage of stream.

File: java-frontend/src/main/java/org/sonar/java/cfg/CFG.java [hunk 3]
```diff
@@ -81,7 +81,13 @@
 
 public class CFG {
 
-  private final Symbol.MethodSymbol methodSymbol;
+  private final boolean ignoreBreakAndContinue;
+
+  public void setMethodSymbol(Symbol.MethodSymbol methodSymbol) {
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/cfg/CFG.java]: Even if [S1213](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1213) is not enabled, I would prefer to have this setter a bit lower in the members declarations (at least after constructors).

**Comment by m-g-sonar** [comment]
There is only indirect tests (tested from the rule) about how you made validation of `Break` and `Continue` statement in CFG optional. As any new feature, it needs proper testing in `CFGTest`, at least to show why it has been added.

File: java-checks/src/main/java/org/sonar/java/checks/ParameterReassignedToCheck.java (hunk 1 above)
Comment (benzonico) [reply java-checks/src/main/java/org/sonar/java/checks/ParameterReassignedToCheck.java]: I see no added value for stream here... neither from readability nor perf...

Comment (m-g-sonar) [reply java-checks/src/main/java/org/sonar/java/checks/ParameterReassignedToCheck.java]: Okay, I was simply wondering about calling 4 times `parameterTree.symbol()`

**Pull request merged**